src/diffusers/pipelines/unclip/pipeline_unclip.py

```diff
@@ -0,0 +1,280 @@
+        clip_extra_context_tokens = self.clip_extra_context_tokens_proj(image_embeddings)
+        clip_extra_context_tokens = clip_extra_context_tokens.reshape(batch_size, -1, self.clip_extra_context_tokens)
+
+        text_encoder_hidden_states = self.text_encoder_hidden_states_proj(text_encoder_hidden_states)
+        text_encoder_hidden_states = self.text_encoder_hidden_states_norm(text_encoder_hidden_states)
+        text_encoder_hidden_states = text_encoder_hidden_states.permute(0, 2, 1)
+        text_encoder_hidden_states = torch.cat([clip_extra_context_tokens, text_encoder_hidden_states], dim=2)
+
+        return text_encoder_hidden_states, additive_clip_time_embeddings
+
+
+class UnCLIPPipeline(DiffusionPipeline):
+    prior: UnCLIPTransformerModel
+    decoder: UNet2DConditionModel
+    embedding_utils: UnCLIPEmbeddingUtils
+    text_encoder: CLIPTextModelWithProjection
+    tokenizer: CLIPTokenizer
+    super_res_64_256_unet_first_steps: UNet2DModel
+    super_res_64_256_unet_last_step: UNet2DModel
+
+    def __init__(
+        self,
+        prior: UnCLIPTransformerModel,
+        decoder: UNet2DConditionModel,
+        text_encoder: CLIPTextModelWithProjection,
+        tokenizer: CLIPTokenizer,
+        embedding_utils: UnCLIPEmbeddingUtils,
+        super_res_64_256_unet_first_steps: UNet2DModel,
+        super_res_64_256_unet_last_step: UNet2DModel,
+    ):
+        super().__init__()
+
+        self.register_modules(
+            prior=prior,
+            decoder=decoder,
+            text_encoder=text_encoder,
+            tokenizer=tokenizer,
+            embedding_utils=embedding_utils,
+            super_res_64_256_unet_first_steps=super_res_64_256_unet_first_steps,
+            super_res_64_256_unet_last_step=super_res_64_256_unet_last_step,
+        )
+
+    def prepare_latents(self, batch_size, num_channels_latents, dtype, device, generator, latents=None):
+        # TODO take as input
+        latent_height = 64
+        latent_width = 64
+
+        shape = (batch_size, num_channels_latents, latent_height, latent_width)
+        if latents is None:
+            if device.type == "mps":
+                # randn does not work reproducibly on mps
+                latents = torch.randn(shape, generator=generator, device="cpu", dtype=dtype).to(device)
+            else:
+                latents = torch.randn(shape, generator=generator, device=device, dtype=dtype)
+        else:
+            if latents.shape != shape:
+                raise ValueError(f"Unexpected latents shape, got {latents.shape}, expected {shape}")
+            latents = latents.to(device)
+
+        # TODO scale latents
+        # scale the initial noise by the standard deviation required by the scheduler
+        # latents = latents * self.scheduler.init_noise_sigma
+        return latents
+
+    @torch.no_grad()
+    @torch.cuda.amp.autocast()  # TODO the original implementation uses autocast
```

Comment:
good to leave for now and for integration testing, we should remove this probs though before merge

Reply:
Sounds good! Was going to ask anyway since I noticed it looks like we don't really use autocast :)

Reply:
as said in the call, let's maybe no test with autocast on as this will make everything less precise